### PR TITLE
Support for MongoDB Replica Set URI with readPreference and host:port enum in MongoDB dictionaries

### PR DESCRIPTION
**Pull request by artem-yadr**
### Changelog category (leave one): 
- Improvement 

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md): 
- Support for connection to a replica set via a URI with a host:port enum and support for the readPreference option in MongoDB dictionaries. Example URI: mongodb://db0.example.com:27017,db1.example.com:27017,db2.example.com:27017/?replicaSet=myRepl&readPreference=primary

### CODE REVIEW

**Review by kssenii — COMMENTED** [review]
Please fix CI checks. See `fast test` failed with "Cannot build ClickHouse". You can see details in `details` button, press `run.log` there.
You will find why build fails:
```
base/poco/MongoDB/CMakeFiles/_poco_mongodb.dir/src/Connection.cpp.o -c /ClickHouse/base/poco/MongoDB/src/Connection.cpp
2023-02-20 15:30:47 /ClickHouse/base/poco/MongoDB/src/Connection.cpp:240:19: error: declaration shadows a local variable [-Werror,-Wshadow]
2023-02-20 15:30:47         Poco::URI theURI(newURI);
2023-02-20 15:30:47                   ^
2023-02-20 15:30:47 /ClickHouse/base/poco/MongoDB/src/Connection.cpp:198:15: note: previous declaration is here
2023-02-20 15:30:47     Poco::URI theURI(newURI);
2023-02-20 15:30:47 
```

**Comment by kssenii** [comment]
```
/ClickHouse/base/poco/MongoDB/src/Connection.cpp
2023-02-27 17:33:54 In file included from ../../ClickHouse/base/poco/MongoDB/src/Connection.cpp:16:
2023-02-27 17:33:54 /ClickHouse/base/poco/MongoDB/include/Poco/MongoDB/Connection.h:165:36: error: return type of out-of-line definition of 'Poco::MongoDB::Connection::uri' differs from that in the declaration
2023-02-27 17:33:54     inline std::string Connection::uri() const
2023-02-27 17:33:54            ~~~~~~~~~~~             ^
2023-02-27 17:33:54 /ClickHouse/base/poco/MongoDB/include/Poco/MongoDB/Connection.h:94:29: note: previous declaration is here
2023-02-27 17:33:54         const std::string & uri() const;
2023-02-27 17:33:54       
```

**Comment by kssenii** [comment]
```
2023-02-28 21:40:20 /ClickHouse/src/Dictionaries/MongoDBDictionarySource.cpp:121:38: error: no member named 'uri' in 'std::shared_ptr<Poco::MongoDB::Connection>'; did you mean to use '->' instead of '.'?
2023-02-28 21:40:20         Poco::URI poco_uri(connection.uri());
2023-02-28 21:40:20                                      ^
2023-02-28 21:40:20                                      ->
2023-02-28 21:40:20 1 error generated.
```

File: base/poco/MongoDB/src/Connection.cpp [hunk 1]
```diff
@@ -145,68 +145,155 @@ void Connection::connect(const Poco::Net::StreamSocket& socket)
...
+        if (socketTimeout > 0)
+        {
+            _socket.setSendTimeout(socketTimeout);
+            _socket.setReceiveTimeout(socketTimeout);
+        }
+        if (strAddresses.size() > 1)
+        {
+            Poco::MongoDB::QueryRequest request("admin.$cmd");
+            request.setNumberToReturn(1);
+            request.selector().add("isMaster", 1);
```
Comment (kssenii) [inline base/poco/MongoDB/src/Connection.cpp:257]: is it correct that here you use `isMaster` and below `ismaster`?

Comment (artem-yadr) [reply base/poco/MongoDB/src/Connection.cpp:257]: > is it correct that here you use `isMaster` and below `ismaster`?


Yes, isMaster is a command that returns the document containing ismaster as a boolean. More about it here: https://www.mongodb.com/docs/v4.0/reference/command/isMaster/

**Review by kssenii — APPROVED** [review]

**Comment by niltonvasques** [comment]
Does it enables MongoDB engine to support uri also? How the connection should be made?

**Pull request merged**